### PR TITLE
LLM Benchmark Decoding Algorithm Rules Proposal.

### PR DESCRIPTION
**Pull request by yyetim**

### CODE REVIEW

File: inference_rules.adoc [hunk 1]
```diff
@@ -821,6 +821,30 @@ The DLRMv2 MLPerf inference code has an option to aggregate multiple consecutive
 2. drawn uniformly from interval [`--samples-to-aggregate-min`, `--samples-to-aggregate-max`]	
 3. drawn from a custom distribution, with its quantile (inverse of CDP) specified in `--samples-to-aggregate-quantile-file=./tools/dist_quantile.txt`.
 
+=== LLM Benchmarks
+
+
+LLM benchmarks contain an auto-regressive generation loop. This loop’s execution is controlled by a decoding algorithm. The following rules discuss the decoding algorithm in the light of MLPerf Inference rules for the Closed submissions:
```
Comment (szutenberg) [inline inference_rules.adoc:827]: This new subsection is in FAQ section. Please keep Q&A format then.

File: inference_rules.adoc [hunk 2]
```diff
@@ -821,6 +821,30 @@ The DLRMv2 MLPerf inference code has an option to aggregate multiple consecutive
...
+
+* Decoding algorithm should follow the implementation available on huggingface: https://huggingface.co/blog/how-to-generate#beam-search
```
Comment (szutenberg) [inline inference_rules.adoc:829]: Please explain beam search algorithm in this document.

[https://huggingface.co/blog/how-to-generate#beam-search ](https://huggingface.co/blog/how-to-generate#beam-search) does not contain any implementation. Please describe how beam search works on a toy example and do not refer to HF docs. We believe that in the inference_rules the algorithm itself should be described, not its implementation.

In HF's beam_search algorithm we can find that topK with k=num_beams*2 is computed: [https://github.com/huggingface/transformers/blob/v4.30.0/src/transformers/generation/utils.py#L2955](https://github.com/huggingface/transformers/blob/v4.30.0/src/transformers/generation/utils.py#L2955). The document "How to generate beam search?" does not mention it and simplifies the algorithm a lot comparing with the actual implementation of beam_search in transformers library. How much freedom does submitter have?

File: inference_rules.adoc [hunk 3]
```diff
@@ -821,6 +821,30 @@ The DLRMv2 MLPerf inference code has an option to aggregate multiple consecutive
 2. drawn uniformly from interval [`--samples-to-aggregate-min`, `--samples-to-aggregate-max`]	
 3. drawn from a custom distribution, with its quantile (inverse of CDP) specified in `--samples-to-aggregate-quantile-file=./tools/dist_quantile.txt`.
 
+=== LLM Benchmarks
+
+
+LLM benchmarks contain an auto-regressive generation loop. This loop’s execution is controlled by a decoding algorithm. The following rules discuss the decoding algorithm in the light of MLPerf Inference rules for the Closed submissions:
+
+* Decoding algorithm should follow the implementation available on huggingface: https://huggingface.co/blog/how-to-generate#beam-search 
+  ** We are using beam=4 for this algorithm’s setting.
```
Comment (szutenberg) [inline inference_rules.adoc:830]: It's better to write that parameters of the decoding algorithm (in [https://github.com/mlcommons/inference/blob/master/language/gpt-j/backend.py#L13](https://github.com/mlcommons/inference/blob/master/language/gpt-j/backend.py#L13) ) must not be changed.

File: inference_rules.adoc [hunk 4]
```diff
@@ -821,6 +821,30 @@ The DLRMv2 MLPerf inference code has an option to aggregate multiple consecutive
...
+  ** The output length determines the runtime of the generation. So, the algorithm in a submission should follow the same settings as the reference model and the algorithm should not change the output length.
```
Comment (szutenberg) [inline inference_rules.adoc:831]: Our proposal: remove “and the algorithm should not change the output length.” -> we just should follow the algorithm

File: inference_rules.adoc [hunk 5]
```diff
@@ -821,6 +821,30 @@ The DLRMv2 MLPerf inference code has an option to aggregate multiple consecutive
...
+* Summary of the allowed decoding algorithm:
```
Comment (szutenberg) [inline inference_rules.adoc:832]: No logical connection with the previous paragraphs, we believe that reference implementation already explains how does the full inference works so no point to duplicate it here.

File: inference_rules.adoc [hunk 6]
```diff
@@ -821,6 +821,30 @@ The DLRMv2 MLPerf inference code has an option to aggregate multiple consecutive
...
+  ** For the LLM benchmarks, each query contains a prompt. Upon receiving the query, the SUT runs this query in two stages: 
+    1. Run the forward pass of the model. As part of this, populate the KV-cache corresponding to this request.
+    2. Start the decoding loop using the beam search algorithm, using 4 beams.
+    3. Loop ends when the max tokens for the benchmark is reached or EOS token is detected as part of beam search.
```
Comment (szutenberg) [inline inference_rules.adoc:836]: Please describe early_stopping=True from HF::
Stop once:
•	Number of generated tokens in all beams reaches max_new_tokens=128 or
•	We found num_beams(=4) candidates.

File: inference_rules.adoc [hunk 7]
```diff
@@ -821,6 +821,30 @@ The DLRMv2 MLPerf inference code has an option to aggregate multiple consecutive
...
+  ** MLPerf inference rules only allow batching as a change of the execution of the model. Following this, the two-stage processing of the query above can execute for a batch of requests in the SUT.
+* Caching details
+  ** As explained earlier about caching, MLPerf Inference rules explicitly disallow caching of queries. KV-cache is a common technique used in caching the state of iterations of a single query in auto-regressive decoding. So, using a KV-cache, which doesn’t apply across queries, is allowed.
+  ** Entries of the KV-cache should be treated similar to activations of a forward pass. They can be quantized with respect to the quantization rules, and following the rules they can’t be pruned.
+  ** A KV-cache row is a tensor that deals with the execution of a single inference query. 
+  ** The entries corresponding to the query need to be reset and cannot retain state across new invocations.
```
Comment (szutenberg) [inline inference_rules.adoc:842]: Reset is very ambiguous term. Our proposal: KV-cache values must not be reused between any two inference calls (inference_call as in the reference implementation).

File: inference_rules.adoc [hunk 8]
```diff
@@ -821,6 +821,30 @@ The DLRMv2 MLPerf inference code has an option to aggregate multiple consecutive
 2. drawn uniformly from interval [`--samples-to-aggregate-min`, `--samples-to-aggregate-max`]	
 3. drawn from a custom distribution, with its quantile (inverse of CDP) specified in `--samples-to-aggregate-quantile-file=./tools/dist_quantile.txt`.
 
+=== LLM Benchmarks
+
+
+LLM benchmarks contain an auto-regressive generation loop. This loop’s execution is controlled by a decoding algorithm. The following rules discuss the decoding algorithm in the light of MLPerf Inference rules for the Closed submissions:
+
```
Comment (szutenberg) [inline inference_rules.adoc:828]: In the whole PR:
does "should" mean "should" from [https://www.ietf.org/rfc/rfc2119.txt](https://www.ietf.org/rfc/rfc2119.txt) or "must"?

**Review by szutenberg — CHANGES_REQUESTED** [review]
Sorry, I won't be able to attend today meeting.

**Comment by yyetim** [comment]
Thanks, I incorporated the changes from the LLM task force meeting, and also from the comments in the PR. PTAL.

We discussed in the task force meeting that Bhadri and Thomas were going to add more details to the beam search reference in the GPT-J. I added a TODO for them.

File: inference_rules.adoc [hunk 9]
```diff
@@ -821,6 +821,20 @@ The DLRMv2 MLPerf inference code has an option to aggregate multiple consecutive
...
 
+=== LLM Benchmarks
+
+Q: What algorithm is used for the auto-regressive decoding loop?
+
+A: The benchmark uses the beam search algorithm described at a high level here: https://huggingface.co/blog/how-to-generate#beam-search. Specifically, we use a beam width of 4 and enable early termination. TODO(thomas,bhadri): Add more details here of the beam search in GPT-J.
+
+Q: MLPerf disallows caching queries. Is using a KV-cache in decoding allowed?
+
+A: Using a KV-cache is allowed in the same way as included in the reference model where it does not apply across queries. A KV-cache row is a tensor that must deal with the execution of a single inference query. Every input query must be computed in its entirety. Entries of the KV-cache must be treated similar to activations of a forward pass. They can be quantized with respect to the quantization rules. Similarly, they cannot be pruned (or sparsified).
```
Comment (akhilarunkumar-dm) [inline inference_rules.adoc:832]: There was concern regarding the "cannot be pruned (or sparsified)" part of this in the inference working group today. Anton brought up a valid point that some implementations might choose to cache some entries while rematerializing some K/V vectors. I think that is a valid scenario that we want to allow. In my mind, we were not trying to prohibit that with the above rule.

Comment (yyetim) [reply inference_rules.adoc]: Thanks for the note Akhil. Yes, I agree. We want to ensure with the rules that we standardize using the KV-cache as it has a lot of moving parts. However, not using a KV cache or rematerializing some of the values should be allowed. Pruning and sparsification changes model predictions and are disallowed in general as part of model equivalence rules. Not using KV-cache isn't pruning as long as the values are rematerialized.

**Review by mrasquinha-g — APPROVED** [review]

**Comment by nv-ananjappa** [comment]
@yyetim Could you sign the MLCommons CLA so that this PR can be merged? CC: @mrasquinha-g

**Comment by yyetim** [comment]
Will do by tomorrow

**Comment by yyetim** [comment]
I submitted the form for the MLCommons CLA, @mrasquinha-g can you help with approval?

**Comment by yyetim** [comment]
Requested access and added the google email used in the commit to my account. Will retry a dummy commit to rerun the check.

**Comment by yyetim** [comment]
OK, the cla-bot check now passed. @mrasquinha-g can you approve again to merge? Thanks

**Review by mrasquinha-g — APPROVED** [review]

**Pull request merged**